### PR TITLE
EZP-29947: Added search criteria tags

### PR DESCRIPTION
**Pull request by lucasOsti**
Question      | Answer
| ------------- | ---
| Tickets       | https://jira.ez.no/browse/EZP-29947, https://jira.ez.no/browse/EZP-29951
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Tests pass?   | yes
| Doc needed?   | no
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

This PR i releted with 2 jira tickets:
- https://jira.ez.no/browse/EZP-29947 added search criteria tags
- https://jira.ez.no/browse/EZP-29951 added posibilty to remove search criteria

#### Checklist:
- [x] Coding standards (`$ composer fix-cs`)
- [x] Ready for Code Review

### CODE REVIEW

**Comment by lucasOsti** [comment]
Zrzut ekranu 2020-03-04 o 10 43 00

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 1]
```diff
@@ -2,10 +2,13 @@
     let getUsersTimeout;
     const CLASS_DATE_RANGE = 'ez-filters__range-wrapper';
     const CLASS_VISIBLE_DATE_RANGE = 'ez-filters__range-wrapper--visible';
+    const SEARCH_TAG_SELECTOR = '.ez-tag';
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]:
```suggestion
    const SELECTOR_TAG = '.ez-tag';
```

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 2]
```diff
@@ -2,10 +2,13 @@
...
+    const CLEAR_BTN_SELECTOR = '.ez-tag__remove-btn';
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]:
```suggestion
    const SELECTOR_REMOVE_TAG = '.ez-tag__remove-btn';
```

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 3]
```diff
@@ -274,6 +284,58 @@
...
+        sectionSelect[0].selected = true;
+        removeSearchTag(event);
+    };
+    const clearSubtree = (event) => {
+        doc.querySelector('#search_subtree-content-breadcrumbs').hidden = true;
+        doc.querySelector('.ez-btn--udw-select-location').hidden = false;
+        subtreeInput.value = '';
+        removeSearchTag(event);
+    };
+    const clearLastModified = (event) => {
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]: `clearLastModified` and `clearLastCreated` have much in common, we can extract into a one method to be DRY

File: src/bundle/Resources/views/themes/admin/ui/search/form.html.twig [hunk 4]
```diff
@@ -19,6 +22,74 @@
...
     </span>
 </div>
+
+<div class="ez-search-criteria-tags ml-4 {% if filters_expanded %} ez-search-criteria-tags--collapsed{% endif %}">
+    {% if form.content_types.vars.data is not empty %}
+        {% for contentType in form.content_types.vars.data %}
+            {{ include('@ezdesign/ui/search_tag.html.twig', {
+                'content': contentType.name,
+                'title': "#{'search.content.type'|trans|desc('Content Type:')} #{contentType.name}",
+                'reletedFiledId': "search_content_types_#{contentType.identifier}",
```
Comment (dew326) [inline src/bundle/Resources/views/themes/admin/ui/search/form.html.twig]: Maybe we could do `targetSelector/fieldSelector` and pass the whole selector not only the id?

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 5]
```diff
@@ -286,6 +340,14 @@
         sectionSelect.addEventListener('change', toggleDisabledStateOnApplyBtn, false);
     }
 
+    for (let tagType in clearSearchTagBtnMethods) {
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]:
```suggestion
    for (const tagType in clearSearchTagBtnMethods) {
```

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 6]
```diff
@@ -286,6 +340,14 @@
...
+        let listenerTarget = doc.querySelectorAll(`.ez-tag__remove-btn--${tagType}`);
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]:
```suggestion
        const tagBtns = doc.querySelectorAll(`.ez-tag__remove-btn--${tagType}`);
```

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 7]
```diff
@@ -274,6 +278,56 @@
...
+    const clearSearchTagBtnMethods = {
+        section: (event) => clearSection(event),
+        subtree: (event) => clearSubtree(event),
+        creator: (event) => clearCreator(event),
+        'content-types': (event) => clearContentType(event),
+        'last-modified': (event) => clearDataRange(event, lastModifiedSelect.dataset.targetSelector),
+        'last-created': (event) => clearDataRange(event, lastCreatedSelect.dataset.targetSelector),
+    };
+
+    dateFields.forEach(initFlatPickr);
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]: one init is enough

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 8]
```diff
@@ -274,6 +278,56 @@
             defaultDate,
         });
     };
+    const removeSearchTag = (event) => {
+        const btn = event.target.closest(SELECTOR_REMOVE_TAG);
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]: You don't need this.

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 9]
```diff
@@ -286,6 +340,14 @@
         sectionSelect.addEventListener('change', toggleDisabledStateOnApplyBtn, false);
     }
 
+    for (let tagType in clearSearchTagBtnMethods) {
+        let listenerTarget = doc.querySelectorAll(`.ez-tag__remove-btn--${tagType}`);
+
+        if (listenerTarget) {
```
Comment (dew326) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]:
```suggestion
        if (listenerTarget.length) {
```

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 10]
```diff
@@ -274,6 +277,54 @@
...
+        endDateInput.vaue = '';
+        dataRange.classList.remove(CLASS_VISIBLE_DATE_RANGE);
+        removeSearchTag(event);
+    };
+    const clearCreator = (event) => {
+        handleResetUser();
+        removeSearchTag(event);
+    };
+    const clearSearchTagBtnMethods = {
+        section: (event) => clearSection(event),
```
Comment (GrabowskiM) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]:
```suggestion
        section: clearSection,
        subtree: clearSubtree,
        creator: clearCreator,
        'content-types': clearContentType,
```

File: src/bundle/Resources/public/js/scripts/admin.search.filters.js [hunk 11]
```diff
@@ -286,6 +337,14 @@
         sectionSelect.addEventListener('change', toggleDisabledStateOnApplyBtn, false);
     }
 
+    for (const tagType in clearSearchTagBtnMethods) {
+        const tagBtns = doc.querySelectorAll(`.ez-tag__remove-btn--${tagType}`);
+
+        if (tagBtns.length) {
```
Comment (GrabowskiM) [inline src/bundle/Resources/public/js/scripts/admin.search.filters.js]: `if` is unnecessary, `querySelectorAll` will always return array and `forEach` can be executed on empty array

File: src/bundle/Resources/views/themes/admin/ui/search_tag.html.twig [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    <div class="ez-tag__content">
+        {{ content }}
+    </div>
+    <button 
+        type="button" 
+        class="ez-tag__remove-btn {{ btnClass }}"
+        data-target-selector="{{ targetSelector }}"
+    >
+        <svg class="ez-icon ez-icon--medium ez-icon--dark">
+            <use xlink:href="/bundles/ezplatformadminui/img/ez-icons.svg#circle-close"></use>
```
Comment (GrabowskiM) [inline src/bundle/Resources/views/themes/admin/ui/search_tag.html.twig]: asset()

Comment (bogusez) [reply src/bundle/Resources/views/themes/admin/ui/search_tag.html.twig]: @lucasOsti I've reported 2 issues regarding this PR on task comment in jira.

**Comment by bogusez** [comment]
@lucasOsti I've reported 2 issues regarding this PR on task comment in jira.

File: src/bundle/Resources/views/themes/admin/ui/search/form.html.twig [hunk 13]
```diff
@@ -19,7 +22,75 @@
...
+
+        {{ include('@ezdesign/ui/search_tag.html.twig', {
+            'content': content_breadcrumbs,
+            'title':  "#{'search.subtree'|trans|desc('Subtree:')} #{content_breadcrumbs}",
+            'btnClass': "ez-tag__remove-btn--subtree",
+        }) }}
+    {% endif %}
+
+    {% if form.vars.value.lastModified is not empty %}
+        {% set startDate = form.vars.value.lastModified.start_date|ez_short_date %}
```
Comment (adamwojs) [inline src/bundle/Resources/views/themes/admin/ui/search/form.html.twig]: Correct me if I'm wrong but according to convention identifiers should use underscore as a word separator e.g. https://github.com/ezsystems/ezplatform-admin-ui/pull/1265/files#diff-4f6a06e0b494e350331247a4b1dd2c54R48

File: src/bundle/Resources/views/themes/admin/ui/search_tag.html.twig [hunk 14]
```diff
@@ -0,0 +1,19 @@
+{% set btnClass = btnClass|default('') %}
+{% set content = content|default('') %}
+{% set targetSelector = targetSelector|default('') %}
+{% set title = title|default('') %}
+
+<div class="ez-tag ez-search-criteria-tags__tag" {% if title %} title="{{ title }} {% endif %}">
```
[the remark this answers is not in the document]
Comment (dew326) [reply src/bundle/Resources/views/themes/admin/ui/search_tag.html.twig]:
```suggestion
<div class="ez-tag ez-search-criteria-tags__tag"{% if title %} title="{{ title }}{% endif %}">
```
If we are so picky, I think it should be like this ;)

**Pull request merged**